rust/gateway/src/main.rs

```diff
@@ -80,3 +98,12 @@ struct Cli {
     #[command(flatten)]
     common: CommonArgs,
 }
+
+/// Checks whether the given [`std::error::Error`] is in-fact an HTTP error with a 4xx status code.
+fn is_client_error(e: &(dyn std::error::Error + 'static)) -> bool {
+    let Some(tungstenite::Error::Http(r)) = e.downcast_ref() else {
```

Comment:
```suggestion
    let Some(phoenix_channel::Error::WebSocket(tungstenite::Error::Http(r))) = e.downcast_ref() else {
```

Seems like we wrapped `tungstenite`'s error with one from `phoenix_channel`, so this downcast failed and we were returning a `Transient` error. 

I tested with this change and it works

Reply:
Hmm, it looks like we are not exposing the error correctly as `source()` then.

Thanks for the pointers in testing this, will iterate on it!